### PR TITLE
feat(voice): add xAI realtime provider

### LINKED ISSUE #16497: [INTEGRATION] xAI Grok Voice Agent API provider
### Service Information

- Service Name: xAI Grok Voice Agent API
- Service Website: https://x.ai/api/voice
- API Documentation: https://docs.x.ai/developers/model-capabilities/audio/voice-agent?campaign=think-fast-blog

### Use Case

Mastra already has the primitives needed for realtime voice providers:

- `MastraVoice` defines the common provider contract.
- `CompositeVoice` supports a `realtime` provider alongside separate input/output providers.
- Agents already accept a `voice` provider through agent config.
- Existing packages such as `@mastra/voice-openai-realtime`, `@mastra/voice-google-gemini-live`, and `@mastra/voice-aws-nova-sonic` establish the package and docs pattern for realtime voice integrations.

xAI now provides a realtime Grok Voice Agent API for bidirectional audio/text over WebSocket. Adding a first-class Mastra provider would let users build Grok-powered voice assistants, phone agents, and interactive voice systems without creating custom WebSocket glue code.

### Proposed Solution

Add a new standalone voice provider package:

- Package directory: `voice/xai-realtime-api`
- Package name: `@mastra/voice-xai-realtime`
- Main class: `XAIRealtimeVoice extends MastraVoice`

The implementation should use `@mastra/voice-openai-realtime` as the closest structural reference, but it should keep xAI-specific protocol behavior explicit instead of treating the OpenAI provider as a drop-in replacement.

Provider defaults from the xAI docs:

- WebSocket URL: `wss://api.x.ai/v1/realtime`
- Default model: `grok-voice-think-fast-1.0`
- Default voice: `eve`
- Other built-in voices: `ara`, `rex`, `sal`, `leo`
- Default audio format: PCM at 24kHz
- Server-side auth: `Authorization: Bearer ${XAI_API_KEY}`

### Expected Capabilities

The provider should support:

- `connect()` for creating the xAI realtime WebSocket session.
- `updateConfig()` via `session.update`.
- `addInstructions()` mapped to xAI session instructions.
- `addTools()` for Mastra function tools.
- `speak()` for text turns using `conversation.item.create` and `response.create`.
- `send()` for realtime audio input using `input_audio_buffer.append`.
- Manual turn control with `input_audio_buffer.commit` and `input_audio_buffer.clear` where applicable.
- `answer()` to trigger `response.create`.
- `close()` / disconnect cleanup.
- `getSpeakers()` returning documented xAI voices.
- `getListener()` returning realtime listener capability.
- Mastra voice events for streamed output:
  - `speaker`
  - `speaking`
  - `speaking.done`
  - `writing`
  - `error`

### xAI-Specific Protocol Handling

The provider should explicitly support xAI’s documented event behavior, including:

- `response.output_audio.delta` for streamed audio output.
- xAI text delta events, including the documented OpenAI compatibility difference around text event naming.
- `response.function_call_arguments.done` for client-side function tools.
- Parallel tool calls: collect all function calls for a response, submit a `function_call_output` item for each one, and send only one continuation `response.create` after all outputs are sent.
- xAI server-side tool passthrough for documented tool types:
  - `file_search`
  - `web_search`
  - `x_search`
  - `mcp`
  - `function`

### Production Requirements

This should be production-ready, not just a demo wrapper:

- Do not leak API keys in logs, spans, or serialized provider state.
- Propagate Mastra `requestContext` into tool execution.
- Map xAI server error frames and WebSocket errors to Mastra `error` events.
- Handle WebSocket close events and document reconnect behavior.
- Avoid overlapping audio after tool calls by allowing continuation timing to be controlled safely.
- Use typed xAI protocol definitions instead of importing OpenAI realtime types blindly.
- Document browser auth separately: browser WebSockets cannot send headers, so users need xAI ephemeral-token/subprotocol auth for client-side usage.
- Document Node/server auth through `XAI_API_KEY`.

### Acceptance Criteria

- A new `@mastra/voice-xai-realtime` package is added under `voice/xai-realtime-api`.
- `XAIRealtimeVoice` can be used directly as an agent voice provider.
- `XAIRealtimeVoice` can be used with `CompositeVoice({ realtime: new XAIRealtimeVoice(...) })`.
- The provider supports xAI session config, text input, audio input, streamed audio/text output, function tools, and documented xAI server-side tools.
- Unit tests cover:
  - default config
  - auth header construction
  - session update payloads
  - text turn creation
  - audio buffer append/commit/clear
  - speaker list
  - event mapping
  - error mapping
  - Mastra tool transformation
  - parallel function-call handling
  - no API key leakage
- Docs are added:
  - package README
  - reference page under `docs/src/content/en/reference/voice/`
  - voice overview entry if appropriate
- A changeset is included.

### Alternatives Considered

The existing `@mastra/voice-openai-realtime` package is close structurally because xAI documents OpenAI Realtime compatibility. However, using it directly would not be production-ready because xAI has provider-specific endpoint/auth, model defaults, voices, server-side tools, event naming differences, and function-call flow requirements.

A dedicated provider keeps the integration explicit while preserving Mastra’s existing voice abstraction.

### Verification

- [x] I have searched the existing issues to make sure this integration hasn't already been requested
- [x] I have checked that this service has a public API that can be integrated

### PR DESCRIPTION
**Pull request by mbenhamd**
## Description

Adds `@mastra/voice-xai-realtime`, a realtime voice provider for the xAI Grok Voice Agent API.

The provider follows Mastra voice primitives and handles xAI WebSocket auth, session updates, text turns, realtime audio streaming, manual buffer controls, response cancellation, built-in xAI voices, Mastra function tools, and xAI server-side tools.

Example:

```ts
import { Agent } from "@mastra/core/agent";
import { XAIRealtimeVoice } from "@mastra/voice-xai-realtime";

const voice = new XAIRealtimeVoice({
  apiKey: process.env.XAI_API_KEY,
  model: "grok-voice-think-fast-1.0",
  speaker: "eve",
  turnDetection: { type: "server_vad" },
});

const agent = new Agent({
  id: "voice-agent",
  name: "Voice Agent",
  instructions: "You are a helpful voice assistant.",
  model: "xai/grok-4.3",
  voice,
});

await agent.voice.connect();
await agent.voice.speak("How can I help you today?");
```

This includes production hardening for reconnects, stale socket events, malformed function-call arguments, missing function-call argument timeouts, parallel tool calls, void tool outputs, request context cleanup, and declaration-file emission for package consumers.

Docs, README, sidebar entry, voice overview entry, package tests, and a changeset are included.

## Related issue(s)

Fixes #16497.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation update
- [ ] Code refactoring
- [ ] Performance improvement
- [x] Test update

## Checklist

- [x] I have linked the related issue(s) in the description above
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have addressed all Coderabbit comments on this PR

## Verification

- `corepack pnpm --filter @mastra/voice-xai-realtime test`
- `corepack pnpm --filter @mastra/voice-xai-realtime lint`
- `corepack pnpm --filter @mastra/voice-xai-realtime build`
- `node_modules/.bin/tsc -p voice/xai-realtime-api/tsconfig.json --noEmit`
- `node_modules/.bin/tsc -p voice/xai-realtime-api/tsconfig.build.json --listEmittedFiles --pretty false`
- `corepack pnpm prettier --check .` from `docs/`
- `corepack pnpm lint:remark` from `docs/`
- `corepack pnpm --filter mastra-docs validate`
- `git diff --check`
- `coderabbit review --plain --base main --no-color`
- `claude -p` focused review
- `codex review --base main`

## Notes

- The provider is built for Node.js server-side runtimes.
- `ephemeralToken` is supported for already-minted xAI tokens and is sent via the `xai-client-secret.<token>` WebSocket protocol instead of an authorization header.
- xAI server-side tools (`web_search`, `x_search`, `file_search`, and `mcp`) are passed through in `session.update`; Mastra function tools are converted to xAI function tools and executed client-side by the provider.

## ELI5 Explanation

This PR adds a new tool to Mastra that lets applications have voice conversations with xAI's Grok AI using WebSocket connections. It's like adding a phone operator that can both listen to what you say and talk back to you in a smooth, two-way conversation.

---

## Overview

This PR introduces `@mastra/voice-xai-realtime`, a new Mastra voice provider package that implements full integration with xAI's realtime WebSocket API (Grok Voice Agent). The provider enables bidirectional text and audio communication, tool execution, and leverages xAI's server-side capabilities including web search and MCP tools.

---

## New Package: `@mastra/voice-xai-realtime`

A complete new npm package has been added with the following structure and features:

**Core Implementation (`src/index.ts`)**
- `XAIRealtimeVoice` class extending Mastra's voice primitives
- Full WebSocket lifecycle management (connect, disconnect, close)
- Mastra voice operations: `speak()`, `listen()`, `send()`, `answer()`, `addInstructions()`, `addTools()`, `cancelResponse()`
- Audio buffer operations: `commitAudioBuffer()`, `clearAudioBuffer()`
- Session configuration: `updateConfig()` with turn detection and audio settings
- Speaker/listener information retrieval
- Request context propagation through tool execution
- Event mapping from xAI protocol to Mastra voice events (speaker, speaking, speaking.done, writing, error)

**Authentication**
- Server-side: Bearer token via `Authorization` header (XAI_API_KEY)
- Browser/ephemeral: Subprotocol-based ephemeral token support via `xai-client-secret.<token>`
- Configurable via `apiKey` or `ephemeralToken` parameters with proper precedence handling

**Audio Support**
- Default: 24 kHz PCM16 format
- Optional telephony codecs: 8 kHz G.711 (µ-law/a-law)
- Configurable input/output audio formats
- Audio delta decoding from base64 to readable streams

**Tool Support**
- Mastra function tools converted to xAI function tools via `transformTools()`
- Tool execution with request context propagation
- Parallel function-call collection and sequencing
- Passthrough and merge of xAI server-side tools (`web_search`, `file_search`, `x_search`, `mcp`)
- Proper handling of tool outputs, errors, and continuation timing

**Production Hardening**
- API key redaction in diagnostic outputs
- Stale event filtering after reconnects
- Malformed/missing function-call argument handling
- Circular reference handling in session serialization
- Proper socket close/reconnect semantics
- Undefined field stripping for payload serialization

---

## Types & Utilities (`src/types.ts`, `src/utils.ts`)

**Type Definitions**
- `XAIRealtimeModel`, `XAIVoice`: Model and speaker string unions
- `XAIAudioConfig`, `XAIAudioFormat`, `XAIAudioFormatType`: Audio configuration types
- `XAITurnDetection`: Server-side voice activity detection parameters
- `XAIFunctionTool`, `XAIServerTool` (`XAIFileSearchTool`, `XAIWebSearchTool`, `XAIXSearchTool`, `XAIMCPTool`): Tool definitions
- `XAISessionConfig`, `XAIRealtimeVoiceConfig`: Configuration interfaces
- `XAIRealtimeSpeakOptions`, `XAIRealtimeListenOptions`, `XAIRealtimeAnswerOptions`: Operation options
- `XAIRealtimeEventMap`: Strongly-typed Mastra event mappings
- `XAISpeaker`: Speaker metadata (name, description, category)
- `XAIClientEvent`, `XAIServerEvent`: WebSocket message contracts

**Utility Functions**
- `isReadableStream()`: Type-guard for Node.js streams
- `int16ArrayToBase64()`: PCM16 encoding for audio transmission
- `readableToBuffer()`, `readableToBase64()`: Stream conversion utilities
- `transformTools()`: Converts Mastra tools to xAI function tools with schema transformation

---

## Documentation

**Reference Documentation (`docs/src/content/en/reference/voice/xai-realtime.mdx`)**
- Complete usage example with agent setup, connection, event handling, and audio streaming
- Constructor and options reference with `PropertiesTable` rendering
- Authentication behavior documentation
- Method references for all public APIs
- Mastra event mapping documentation
- Tool support explanation (conversion and passthrough)
- Supported audio formats and sample rates

**Voice Overview (`docs/src/content/en/docs/voice/overview.mdx`)**
- Added xAI Realtime provider to packages list
- Speech-to-speech example with real-time audio streaming
- Configuration example with API key, model, speaker, VAD settings, and server tools
- Reference link integration

**README (`voice/xai-realtime-api/README.md`)**
- Installation and setup instructions
- `XAI_API_KEY` configuration for server-side auth
- Ephemeral token guidance for browser scenarios
- Complete TypeScript usage example
- Server-side tool documentation (web_search, x_search, file_search, mcp)
- Mastra tool conversion and tool call execution explanation
- Supported speakers and custom IDs
- Audio format configuration guide

**Sidebar Entry (`docs/src/content/en/reference/sidebars.js`)**
- Added `voice/xai-realtime` reference documentation link under Voice category

---

## Configuration & Build

**Package Configuration (`package.json`)**
- ESM/CJS dual exports via `exports` field with shared type declarations
- Workspace dependencies: `@mastra/schema-compat`, `ws`
- Dev dependencies including Mastra core and Zod from catalogs
- Peer dependencies: `@mastra/core`, `zod`
- Build, test, and lint scripts
- `prepack` step for package documentation generation
- Node engine requirement: `>=22.13.0`

**Build Configuration**
- `tsup.config.ts`: ESM/CJS bundling with code splitting, treeshake, sourcemaps, and post-build type generation
- `tsconfig.build.json`: Build-specific configuration excluding tests
- `tsconfig.json`: Source tree compilation with test exclusions

**Linting & Formatting**
- `eslint.config.js`: Top-level await configuration with JSDoc type annotation
- `lint-staged.config.js`: Git pre-commit hooks for TypeScript/JavaScript linting and formatting, with JSON/Markdown/YAML formatting

---

## Testing

**Comprehensive Test Suite (`src/index.test.ts`)**
- Mocked WebSocket implementation with auto-open and event simulation
- Default initialization (speakers, listener)
- Connection URL construction and authentication (API key, ephemeral token, error cases)
- Event queuing and ordering semantics (initial session.update before speak, concurrent connect deduplication)
- Pre-connection error handling (audio send before open)
- Audio buffer operations (append, commit, clear for Int16Array)
- Response cancellation and lifecycle
- Listen behavior and response.create triggering
- Realtime event mapping (speaking, writing, done, error)
- Speaker stream lifecycle and text-only response handling
- Tool integration: conversion, parallel execution, out-of-order response.done handling, argument timeout, parse diagnostics, undefined result serialization
- Tool error handling with deferred continuation
- Lifecycle resilience (close/reconnect semantics, stale event filtering, circular reference handling)
- Request context cleanup and propagation
- Instruction persistence across reconnects
- Error emission on post-close operations
- ServerTools merging
- API key redaction in diagnostic outputs

**Utility Tests (`src/utils.test.ts`)**
- PCM16 Int16Array to base64 encoding
- Node readable-stream type-guarding
- Readable-stream to base64 conversion
- Tool transformation (Zod schema conversion to xAI JSON shape)
- Tool invocation with requestContext and toolCallId forwarding
- Skipping tools without execute function

---

## Changeset

A `minor` release changeset has been added documenting the introduction of `XAIRealtimeVoice` with usage guidance showing instantiation, connection, and audio streaming with Mastra Agent.

[Review Change Stack](https://app.coderabbit.ai/change-stack/mastra-ai/mastra/pull/16507)

### CODE REVIEW

File: docs/src/content/en/docs/voice/overview.mdx [hunk 1]
```diff
@@ -629,6 +630,49 @@ For detailed configuration options and advanced features, check out [Speech to S
     Visit the [OpenAI Voice Reference](/reference/voice/openai-realtime) for more information on the OpenAI voice provider.
   </TabItem>
 
+  <TabItem value="xai" label="xAI">
```
Comment (wardpeet) [inline docs/src/content/en/docs/voice/overview.mdx]: shouldn't this be alphabetic?

File: voice/xai-realtime-api/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,1016 @@
...
+  XAIRealtimeVoiceConfig,
+  XAIServerEvent,
+  XAISessionConfig,
+  XAISpeaker,
+  XAITurnDetection,
+  XAITool,
+  XAIVoice,
+} from './types';
+import type { XAIExecuteFunction, XAITransformedTool } from './utils';
+import { int16ArrayToBase64, isReadableStream, readableToBase64, transformTools } from './utils';
```
Comment (wardpeet) [inline voice/xai-realtime-api/src/index.ts:24]: Minor: `getOrCreateSpeakerStream` is a thin wrapper around `getOrCreateStream` with no additional logic. Consider inlining it at the two call sites to reduce indirection — `this.getOrCreateStream(this.speakerStreams, id)` is already clear enough on its own.

File: voice/xai-realtime-api/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,1016 @@
...
+    if (!value || typeof value !== 'object') {
+      return value;
+    }
+
+    if (depth >= maxDepth) {
+      throw new Error('Cannot serialize xAI realtime event: maximum object depth exceeded');
+    }
+
+    const objectValue = value as object;
+    if (Buffer.isBuffer(value) || ArrayBuffer.isView(value) || value instanceof ArrayBuffer) {
```
Comment (wardpeet) [inline voice/xai-realtime-api/src/index.ts:976]: Consider using `this.logger.warn(...)` instead of `console.warn(...)` here for consistency with the rest of the class and to respect the configured log level. Also, it would be helpful to include the tool name in the log message so users can identify which tool was skipped, e.g.:

```ts
this.logger.warn(
  `Skipping tool "${tool.name}" — tools must have a JSON schema defined to be used with the realtime API`,
);
```

**Review by wardpeet — APPROVED** [review]
Great work — production-grade edge case handling, excellent test coverage, and follows monorepo conventions throughout. Left two minor suggestions (inline `getOrCreateSpeakerStream`, use `this.logger.warn` in `transformTools`) but nothing blocking. Nice addition.

File: docs/src/content/en/docs/voice/overview.mdx (hunk 1 above)
Comment (mbenhamd) [reply docs/src/content/en/docs/voice/overview.mdx]: Moved xAI to the end of the speech-to-speech tab list in d76de3a8cc. Order is now OpenAI → Google → AWS Nova Sonic → xAI, so the new entry slots in alphabetically relative to the existing tabs without reshuffling the rest. The other tab groups in this file aren't alphabetic either; happy to do a file-wide pass in a follow-up if you'd like.

File: voice/xai-realtime-api/src/index.ts (hunk 2 above)
Comment (mbenhamd) [reply voice/xai-realtime-api/src/index.ts:24]: Inlined in 3a395a73cd. The wrapper just delegated to `createSpeakerStream`, which already does the get-or-create — only one caller left, switched to `this.createSpeakerStream(responseId)` and dropped the method.

File: voice/xai-realtime-api/src/index.ts (hunk 3 above)
Comment (mbenhamd) [reply voice/xai-realtime-api/src/index.ts:976]: Switched in 3a395a73cd. `transformTools` now takes an optional logger (`{ warn(msg) }`); the class call site forwards `this.logger` so warnings honor the configured Mastra log level. Default falls back to `console` for direct callers. Tool name is already in the message (`Skipping xAI realtime tool "<name>" because…`); tests cover both the logger and the console paths.

**Review by wardpeet — APPROVED** [review]
